ssz/sedes/serializable.py

```diff
@@ -0,0 +1,419 @@
+import abc
+import collections
+import copy
+import re
+
+from eth_utils import (
+    to_dict,
+    to_set,
+    to_tuple,
+)
+
+from ssz.constants import (
+    CONTAINER_PREFIX_LENGTH,
+)
+from ssz.exceptions import (
+    DeserializationError,
+    SerializationError,
+)
+
+
+class MetaBase:
+    fields = None
+    field_names = None
+    field_attrs = None
+
+
+def _get_duplicates(values):
+    counts = collections.Counter(values)
+    return tuple(
+        item
+        for item, num in counts.items()
+        if num > 1
+    )
+
+
+def validate_args_and_kwargs(args, kwargs, arg_names, allow_missing=False):
+    duplicate_arg_names = _get_duplicates(arg_names)
+    if duplicate_arg_names:
+        raise TypeError("Duplicate argument names: {0}".format(sorted(duplicate_arg_names)))
+
+    needed_kwargs = arg_names[len(args):]
+    used_kwargs = set(arg_names[:len(args)])
+
+    duplicate_kwargs = used_kwargs.intersection(kwargs.keys())
+    if duplicate_kwargs:
+        raise TypeError("Duplicate kwargs: {0}".format(sorted(duplicate_kwargs)))
+
+    unknown_kwargs = set(kwargs.keys()).difference(arg_names)
+    if unknown_kwargs:
+        raise TypeError("Unknown kwargs: {0}".format(sorted(unknown_kwargs)))
+
+    missing_kwargs = set(needed_kwargs).difference(kwargs.keys())
+    if not allow_missing and missing_kwargs:
+        raise TypeError("Missing kwargs: {0}".format(sorted(missing_kwargs)))
+
+
+@to_tuple
+def merge_kwargs_to_args(args, kwargs, arg_names, allow_missing=False):
+    validate_args_and_kwargs(args, kwargs, arg_names, allow_missing=allow_missing)
+
+    needed_kwargs = arg_names[len(args):]
+
+    yield from args
+    for arg_name in needed_kwargs:
+        yield kwargs[arg_name]
+
+
+@to_dict
+def merge_args_to_kwargs(args, kwargs, arg_names, allow_missing=False):
+    validate_args_and_kwargs(args, kwargs, arg_names, allow_missing=allow_missing)
+
+    yield from kwargs.items()
+    for value, name in zip(args, arg_names):
+        yield name, value
+
+
+class BaseSerializable(collections.Sequence):
+    def __init__(self, *args, **kwargs):
+        if kwargs:
+            field_values = merge_kwargs_to_args(args, kwargs, self._meta.field_names)
+        else:
```

Comment:
I think we should be able to get rid of this, the `merge_...` functions should be able to handle empty `kwargs`, right?

Reply:
Yeah, this is actually taken care by `merge_kwargs_to_args`.